File: sop-java-picocli/src/main/java/sop/cli/picocli/commands/ListProfilesCmd.java
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package sop.cli.picocli.commands;

import picocli.CommandLine;
import sop.Profile;
import sop.cli.picocli.Print;
import sop.cli.picocli.SopCLI;
import sop.exception.SOPGPException;
import sop.operation.ListProfiles;

@CommandLine.Command(name = "list-profiles",
        resourceBundle = "msg_list-profiles",
        exitCodeOnInvalidInput = 37)
public class ListProfilesCmd extends AbstractSopCmd {

    @CommandLine.Parameters(paramLabel = "COMMAND", arity = "1", descriptionKey = "subcommand")
    String subcommand;

    @Override
    public void run() {
        ListProfiles listProfiles = throwIfUnsupportedSubcommand(
                SopCLI.getSop().listProfiles(), "list-profiles");

        try {
            for (Profile profile : listProfiles.subcommand(subcommand)) {
                Print.outln(profile.toString());
            }
        } catch (SOPGPException.UnsupportedProfile e) {
            String errorMsg = getMsg("sop.error.feature_support.subcommand_does_not_support_profiles", subcommand);
            throw new SOPGPException.UnsupportedProfile(errorMsg, e);
        }
    }
}

File: sop-java-picocli/src/main/kotlin/sop/cli/picocli/commands/ListProfilesCmd.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package sop.cli.picocli.commands

import picocli.CommandLine.Command
import picocli.CommandLine.Parameters
import sop.cli.picocli.SopCLI
import sop.exception.SOPGPException
import sop.exception.SOPGPException.UnsupportedProfile

@Command(
    name = "list-profiles",
    resourceBundle = "msg_list-profiles",
    exitCodeOnInvalidInput = SOPGPException.UnsupportedOption.EXIT_CODE)
class ListProfilesCmd : AbstractSopCmd() {

    @Parameters(paramLabel = "COMMAND", arity = "1", descriptionKey = "subcommand")
    lateinit var subcommand: String

    override fun run() {
        val listProfiles =
            throwIfUnsupportedSubcommand(SopCLI.getSop().listProfiles(), "list-profiles")

        try {
            listProfiles.subcommand(subcommand).forEach { println(it) }
        } catch (e: UnsupportedProfile) {
            val errorMsg =
                getMsg("sop.error.feature_support.subcommand_does_not_support_profiles", subcommand)
            throw UnsupportedProfile(errorMsg, e)
        }
    }
}
